### PR TITLE
sketching out how to use base

### LINKED ISSUE #1: Add {prefix4} support
This projects needs support for the new `{prefix4}` strategy in tilelive-s3.

https://github.com/mapbox/tilelive-s3/pull/77

### PR DESCRIPTION
**Pull request by dnomadb**
cc @rclark Fixes #1

### CODE REVIEW

**Comment by dnomadb** [comment]
Running:

```
[s3prefixed@master] s3prefixed ls s3://tilestream-tilesets-staging/{prefix}/zignal | sort > /tmp/s3prefixed-at-master.txt 
```

Then

```
[s3prefixed@prephix-phor] s3prefixed ls s3://tilestream-tilesets-staging/{prefix}/zignal | sort > /tmp/s3prefixed-at-prephix-phor.txt 
```

We get

```
diff /tmp/s3prefixed-at-master.txt /tmp/s3prefixed-at-prephix-phor.txt

```

No diff. @rclark instead of testing off live `ls`s, I think I am going to unit test the prefix regex / base getter. Thoughts?

**Comment by rclark** [comment]
Yeah adding some unit tests sounds great. Such a thing could even ... spread to other functions, like some kind of virus.

**Comment by dnomadb** [comment]
@rclark I _think_ this is ready to go - testing `{prefix4}` requests are gnarly b/c of how many are made.

File: index.js [hunk 1]
```diff
@@ -21,10 +45,11 @@ ListStream.prototype._read = function() {
   var stream = this;
   var q = queue();
   var prefix;
+  var prefixer = new Prefixer(stream.key);
 
-  for (var i = 0; i <= 16; i++) {
-    for (var j = 0; j <= 16; j++) {
-      prefix = stream.key.replace('{prefix}', i.toString(16) + j.toString(16));
+  for (var i = 0; i < prefixer.base; i++) {
```
Comment (rclark) [inline index.js:50]: 0 to 16, including 16, is 17 possibilities -- and `16.toString(16)` is not something we want:

```
> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16].map(function(i) {
... return i.toString(16);
... });
[ '0',
  '1',
  '2',
  '3',
  '4',
  '5',
  '6',
  '7',
  '8',
  '9',
  'a',
  'b',
  'c',
  'd',
  'e',
  'f',
  '10' ]
```

Comment (dnomadb) [inline index.js:50]: lol
image

**Comment by dnomadb** [comment]
@willwhite @rclark I am starting to worry about this strategy. `s3prefix ls`ing `{prefix4}` takes a _long_ time - I've waited for 15 minutes w/ nothing coming back. If I set the queue https://github.com/mapbox/s3prefixed/blob/prephix-phor/index.js#L46 to a smaller number, I start getting results faster. Do you think that there is there any change that we are overloading `s3.listObjects` https://github.com/mapbox/s3prefixed/blob/prephix-phor/index.js#L117 ?

**Comment by rclark** [comment]
Setting the queue lower makes a ton of sense -- I should never have let this be `Infinity`. Although the honest truth is that this brute-force approach to listing prefixed objects just doesn't scale to 65,536 very well.

File: index.js (hunk 1 above)
Comment (rclark) [inline index.js:50]: i readed it wrong

**Comment by dnomadb** [comment]
Going to merge + publish. While we don't fully have `{prefix4} sorted out, this fixed a bug and added tests.

**Pull request merged**